app/services/combined_audio_derivative_creator.rb

```diff
@@ -0,0 +1,163 @@
+  def duration_of_audio_file_in_seconds(path)
+    options = ['ffprobe', '-v', 'error',
+      '-show_entries', 'format=duration', '-of',
+      'default=noprint_wrappers=1:nokey=1'
+    ] + [ path ]
+    @cmd.run(*options).out.strip
+  end
+
+  def download_components
+    result = []
+    audio_member_files.each do |original_file|
+      log "Downloading #{original_file.metadata['filename']}"
+      new_temp_file = Tempfile.new(['temp_', original_file.metadata['filename'].downcase], :encoding => 'binary')
+      original_file.open(rewindable:false) do |input_audio_io|
+        new_temp_file.write input_audio_io.read until input_audio_io.eof?
+      end
+      result << new_temp_file
+    end
+    result
+  end
+
+  # Generate ffmpeg command to concatenate a set of audio files using
+  # ffmpeg stream mapping.
+  # This is tricky but documented at
+  # https://trac.ffmpeg.org/wiki/Map .
+  # For now, the audio settings are hardwired.
+  # This does not run the command; it just returns
+  # an array that can be passed to @cmd.run .
+  def args_for_ffmpeg(output_file_path)
+    # ffmpeg -y: overwrite the already-existing temp file
+    ffmpeg_command = ['ffmpeg', '-y']
+
+    # Then a list of input files specified with -i
+    input_files = @components.map {|x| [ "-i", x.path] }.flatten
+
+    # List the number of audio streams: one per file
+    stream_list = 0.upto(@components.count - 1).to_a.map{ |n| "[#{n}:a]"}.join
+
+    # Specify what to do with the audio streams:
+    #
+    # Stream mapping:
+    #   Stream #0:0 (mp3float) -> concat:in0:a0
+    #   Stream #1:0 (mp3float) -> concat:in1:a0
+    #   Stream #2:0 (mp3float) -> concat:in2:a0
+    #   [ and so on]
+    #   concat -> Stream #0:0 (libmp3lame)
+    #
+    # v=0 means there is no video.
+    filtergraph = "concat=n=#{@components.count}:v=0:a=1[a]"
+
+    # Finally, some output options:
+    # -map [a] : map just the audio to the output
+    #
+    # -ac 1 : output should be mono
+    #    see https://trac.ffmpeg.org/wiki/AudioChannelManipulation
+    # -b:a 64K: make the output a constant 64k bitrate
+    #    see https://trac.ffmpeg.org/wiki/Encode/MP3
+    output_quality = ["-map", "[a]", "-ac", "1", "-b:a", "64k"]
+
+    # Put it all together and return:
+    ffmpeg_command +
+      input_files +
+      ['-filter_complex'] +
+      ["#{stream_list}#{filtergraph}"] +
+      output_quality +
+      [output_file_path]
+  end
+
+  def audio_member_files
+    audio_members.map { |asset| asset.file }
+  end
+
+  # Filters out non-audio items
+  def audio_members
+    work.members.order(:position, :id).select do |member|
+      member.stored? && member.file.mime_type.start_with?('audio')
+    end
+  end
+
+  # If this checksum changes, you need to regenerate the audio
+  def fingerprint
+    @calculated_checksum ||= begin
+      parts = [work.title, work.friendlier_id] +
+        audio_members.map { |a| a.file.metadata['sha512'][0..10] }
+      Digest::MD5.hexdigest(parts.join)
```

Comment:
Why are you including only the first 10 chars of the sha512?  While that *might* be safe, it runs some risk, and I don't see any downside of including the whole thing. 

Additionally, when we make `durations`  or `start_times` a pair of id=>time, we could include whatever id we choose there in the fingerprint too, to ensure that if that changes, the fingerprint shows up as stale. That might make it fine to use the easier friendlier_id instead of uuid pk.

Reply:
Also, you can just do `asset.sha512` as a shortcut, don't need to do `asset.file.metadata['sha512']`. The shortcut seems less error prone too. (Should we raise if any of the sha512's are missing to avoid a subsequent harder to debug problem?)

Reply:
`asset.sha512` -- good call. I'm also deferring to you re: using the entire hash, although 40 bits of info per asset seem plenty for our purposes here. (On the other hand, truncating makes things more complicated and bug prone.)